internal/metadata/metadata.go

```diff
@@ -92,7 +92,7 @@ func Validate(md metadata.MD) error {
 		for i := 0; i < len(k); i++ {
 			r := k[i]
 			if !(r >= 'a' && r <= 'z') && !(r >= '0' && r <= '9') && r != '.' && r != '-' && r != '_' {
-				return errors.New("header key contains not 0-9a-z-_. characters")
+				return fmt.Errorf("header key %q contains not 0-9a-z-_. characters", k)
```

Comment:
"contains illegal characters" (probably good enough)
or
"contains characters not in [0-9a-z_.-]"
or something like that